## 실습 1. Excel 파일 불러오기

해야 할 일

pandas를 이용해 Excel 파일을 DataFrame으로 불러옵니다.

먼저 코드부터 작성하기보다 인공지능에게 현재 상황과 원하는 작업을 설명해 봅니다.

In [27]:
import pandas as pd

df = pd.read_excel("교보문고_종합_베스트셀러_상품리스트.xlsx")

df.head()

c:\Dev\llm-data-analysis\.venv\Lib\site-packages\openpyxl\styles\stylesheet.py:237: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")


,순위,상품코드,판매상품 ID,상품명,정가,판매가,할인율,적립율,적립예정포인트,인물,출판사,발행(출시)일자,분야
0,1,9788936434120,S000000610612,소년이 온다,"15,000","13,500",10%,5%,750,한강,창비,20140519,소설
1,2,9788998441012,S000001632467,모순,"13,000","11,700",10%,5%,650,양귀자,쓰다,20130401,소설
2,3,9788997780624,S000216262757,결국 국민이 합니다,"22,000","19,800",10%,5%,"1,100",이재명,오마이북,20250415,정치/사회
3,4,9788936439743,S000216111714,혼모노,"18,000","16,200",10%,5%,900,성해나,창비,20250328,소설
4,5,9788937473401,S000200550189,급류,"14,000","12,600",10%,5%,700,정대건,민음사,20221222,소설


### 실행 결과

- `pandas`를 이용해 Excel 파일을 DataFrame으로 불러왔습니다.
- `head()`를 사용하여 데이터의 상위 5개 행을 확인했습니다.
- 순위, 상품코드, 판매상품 ID, 상품명, 정가 등의 데이터를 확인할 수 있었습니다.

## 실습 2. 데이터의 크기와 컬럼 확인하기

 
Excel 파일을 불러왔으면 바로 전처리를 시작하지 않습니다.

 
먼저 어떤 데이터인지 확인합니다.

In [28]:
print("1. 전체 행과 컬럼 개수")
print(df.shape)

print("\n2. 컬럼 이름")
print(df.columns)

print("\n3. 자료형")
print(df.dtypes)

print("\n4. 결측치 개수")
print(df.isna().sum())

print("\n5. 앞의 5행")
print(df.head())

1. 전체 행과 컬럼 개수
(199, 13)

2. 컬럼 이름
Index(['순위', '상품코드', '판매상품 ID', '상품명', '정가', '판매가', '할인율', '적립율', '적립예정포인트',
       '인물', '출판사', '발행(출시)일자', '분야'],
      dtype='str')

3. 자료형
순위          int64
상품코드        int64
판매상품 ID       str
상품명           str
정가            str
판매가           str
할인율           str
적립율           str
적립예정포인트       str
인물            str
출판사           str
발행(출시)일자    int64
분야            str
dtype: object

4. 결측치 개수
순위          0
상품코드        0
판매상품 ID     0
상품명         0
정가          0
판매가         0
할인율         0
적립율         0
적립예정포인트     0
인물          0
출판사         0
발행(출시)일자    0
분야          0
dtype: int64

5. 앞의 5행
   순위           상품코드        판매상품 ID         상품명      정가     판매가  할인율 적립율  \
0   1  9788936434120  S000000610612      소년이 온다  15,000  13,500  10%  5%   
1   2  9788998441012  S000001632467          모순  13,000  11,700  10%  5%   
2   3  9788997780624  S000216262757  결국 국민이 합니다  22,000  19,800  10%  5%   
3   4  9788936439743  S000216111714         혼모노  18,00

### 실행 결과

- 총 **199행, 13개 컬럼**으로 구성된 데이터입니다.
- `순위`, `상품코드`, `발행(출시)일자`는 정수형이며, 나머지는 대부분 문자열 자료형입니다.
- **결측치는 모든 컬럼에서 발견되지 않았습니다.**
- `head()`를 통해 상위 5개 도서 데이터를 확인했습니다.

## 실습 3. 분석에 필요한 컬럼만 선택하기


원본 데이터의 모든 컬럼이 이후 분석에 필요한 것은 아닙니다.

 
이번 과정에서는 다음 컬럼을 먼저 선택합니다.

In [29]:
df_books = df[
    ['순위', '판매상품 ID', '상품명', '판매가',
     '인물', '출판사', '발행(출시)일자', '분야']
]

df_books.head()

,순위,판매상품 ID,상품명,판매가,인물,출판사,발행(출시)일자,분야
0,1,S000000610612,소년이 온다,"13,500",한강,창비,20140519,소설
1,2,S000001632467,모순,"11,700",양귀자,쓰다,20130401,소설
2,3,S000216262757,결국 국민이 합니다,"19,800",이재명,오마이북,20250415,정치/사회
3,4,S000216111714,혼모노,"16,200",성해나,창비,20250328,소설
4,5,S000200550189,급류,"12,600",정대건,민음사,20221222,소설


### 실행 결과

- 필요한 **8개 컬럼을 선택하여 `df_books` DataFrame을 생성했습니다.**
- `head()`를 통해 상위 5개 도서의 데이터가 정상적으로 선택된 것을 확인했습니다.
- 순위, 상품명, 판매가, 인물, 출판사, 분야 등의 정보를 확인할 수 있습니다.

## 실습 4. 컬럼 이름을 사용하기 쉽게 정리하기
 
원본 컬럼 이름은 사람이 읽기에는 좋지만 코드에서 반복해서 사용하기에는 조금 길 수 있습니다.

 
이번 실습에서는 다음처럼 간단하게 정리할 수 있습니다.

In [30]:
df_books = df_books.rename(columns={
    '판매상품 ID': '판매상품ID',
    '인물': '저자',
    '발행(출시)일자': '발행일'
})

df_books.columns

Index(['순위', '판매상품ID', '상품명', '판매가', '저자', '출판사', '발행일', '분야'], dtype='str')

### 실행 결과

- `rename()`을 이용해 3개의 컬럼 이름을 변경했습니다.
- `판매상품 ID` → `판매상품ID`
- `인물` → `저자`
- `발행(출시)일자` → `발행일`
- `columns`를 통해 변경된 컬럼명을 정상적으로 확인했습니다.

## 실습 5. 결측치 확인하기

결측치는 값이 비어 있는 데이터를 의미합니다.
 
데이터 분석에서는 결측치를 무조건 삭제하지 않습니다.
 
먼저 어디에 얼마나 있는지 확인하고, 분석 목적에 맞게 처리 방법을 결정합니다.

In [31]:
# 결측치 확인
df_books.isna().sum()


# 결측치가 있는 컬럼만
# df_books.isna().sum()[df_books.isna().sum() > 0]

순위        0
판매상품ID    0
상품명       0
판매가       0
저자        0
출판사       0
발행일       0
분야        0
dtype: int64

### 실행 결과

- `df_books`의 모든 컬럼에서 **결측치가 0개**로 확인되었습니다.
- 따라서 현재 데이터에는 별도의 결측치 처리가 필요하지 않습니다.

## 실습 6. 판매가를 숫자로 바꾸기

Excel에서 판매가가 다음과 같이 저장되어 있을 수 있습니다.

In [36]:
# 1. 쉼표 제거
# df_books['판매가'] = df_books['판매가'].str.replace(',', '')
# 2. 숫자형으로 변환
# df_books['판매가'] = pd.to_numeric(df_books['판매가'])
# 3. 자료형 확인
df_books['판매가'].dtype


df_books["판매가"].head()

0    13500
1    11700
2    19800
3    16200
4    12600
Name: 판매가, dtype: int64

### 실행 결과

- `판매가` 컬럼의 쉼표를 제거했습니다.
- 문자열 데이터를 숫자형으로 변환했습니다.
- 변환 후 `dtype`이 **int64**로 확인되었습니다.
- 상위 5개 판매가 데이터도 정상적으로 확인했습니다.

## 실습 7. 발행일을 날짜형으로 바꾸기

원본 데이터의 발행일은 다음처럼 보일 수 있습니다.

In [37]:
# 날짜형으로 변환
df_books['발행일'] = pd.to_datetime(
    df_books['발행일'],
    format='%Y%m%d'
)

# 자료형 확인
print(df_books['발행일'].dtype)

# 앞의 5개 값 확인
print(df_books['발행일'].head())

datetime64[us]
0   2014-05-19
1   2013-04-01
2   2025-04-15
3   2025-03-28
4   2022-12-22
Name: 발행일, dtype: datetime64[us]


### 실행 결과

- `발행일` 컬럼을 `YYYYMMDD` 형식에서 날짜 자료형으로 변환했습니다.
- `dtype`이 **datetime64[us]**로 확인되어 날짜형 변환이 정상적으로 완료되었습니다.
- 앞의 5개 날짜도 정상적으로 확인했습니다.

## 실습 8. 문자열 앞뒤 공백 정리하기
 
상품명, 저자, 출판사, 분야는 문자열 데이터입니다.

문자열 앞뒤에 불필요한 공백이 있으면 같은 값도 서로 다른 데이터처럼 처리될 수 있습니다.

예를 들어 다음 두 값은 사람이 보기에는 같지만 문자열로는 다릅니다.

In [38]:
# 공백을 제거할 컬럼
columns = ['상품명', '저자', '출판사', '분야']

# 각 컬럼의 앞뒤 공백 제거
for col in columns:
    df_books[col] = df_books[col].str.strip()

# 각 컬럼의 앞 5개 값 확인
for col in columns:
    print(f'\n{col}')
    print(df_books[col].head())


상품명
0        소년이 온다
1            모순
2    결국 국민이 합니다
3           혼모노
4            급류
Name: 상품명, dtype: str

저자
0     한강
1    양귀자
2    이재명
3    성해나
4    정대건
Name: 저자, dtype: str

출판사
0      창비
1      쓰다
2    오마이북
3      창비
4     민음사
Name: 출판사, dtype: str

분야
0       소설
1       소설
2    정치/사회
3       소설
4       소설
Name: 분야, dtype: str


### 실행 결과

- `상품명`, `저자`, `출판사`, `분야` 컬럼의 앞뒤 공백을 제거했습니다.
- 각 컬럼의 상위 5개 데이터를 확인하여 정상적으로 처리된 것을 확인했습니다.

## 실습 9. 중복 데이터 확인하기

데이터에서 중복을 발견했다고 해서 바로 삭제하면 안 됩니다.

먼저 무엇을 기준으로 중복인지 생각해야 합니다.

예를 들어 같은 상품명이 여러 번 등장할 수 있습니다.

하지만 같은 제목이라도 다른 판본, 다른 상품, 개정판 등이 있을 수 있습니다.

따라서 이번 데이터에서는 상품명만 보고 중복 행을 바로 삭제하지 않습니다.

우선 상품을 구분하는 판매상품ID를 확인합니다.

In [39]:
# 1. 전체 행이 완전히 같은 중복 개수
print(df_books.duplicated().sum())

# 2. 판매상품ID가 중복된 행 개수
print(df_books.duplicated(subset='판매상품ID').sum())

# 3. 상품명이 중복된 행 개수
print(df_books.duplicated(subset='상품명').sum())

0
0
0


도서 데이터(df_books)에서 중복 여부를 확인한 결과는 다음과 같습니다.
- 전체 컬럼 값이 완전히 동일한 행의 중복 개수: 0개
- 판매상품ID가 중복된 행의 개수: 0개
- 상품명이 중복된 행의 개수: 0개
따라서 현재 데이터에는 완전히 동일한 중복 행이 없으며, 각 도서의 판매상품ID와 상품명도 중복되지 않는 것으로 확인되었습니다.

## 실습 10. 전처리 결과 검증하기

지금까지 여러 전처리를 수행했습니다.

마지막에는 반드시 데이터가 의도한 형태가 되었는지 다시 확인합니다.

In [40]:
# 1. 데이터 크기 확인
print("1. 데이터 크기 (행, 열)")
print(df_books.shape)

# 2. 컬럼명 확인
print("\n2. 컬럼명")
print(df_books.columns)

# 3. 데이터 타입 확인
print("\n3. 데이터 타입")
print(df_books.dtypes)

# 4. 결측치 개수 확인
print("\n4. 컬럼별 결측치 개수")
print(df_books.isna().sum())

# 5. 판매가 앞의 5개 값 확인
print("\n5. 판매가 앞의 5개 값")
print(df_books["판매가"].head())

# 6. 발행일 앞의 5개 값 확인
print("\n6. 발행일 앞의 5개 값")
print(df_books["발행일"].head())

# 7. 판매상품ID 중복 개수 확인
print("\n7. 판매상품ID 중복 개수")
print(df_books.duplicated(subset="판매상품ID").sum())

# 8. 데이터 앞부분 확인
print("\n8. 데이터 앞의 5개 행")
print(df_books.head())

1. 데이터 크기 (행, 열)
(199, 8)

2. 컬럼명
Index(['순위', '판매상품ID', '상품명', '판매가', '저자', '출판사', '발행일', '분야'], dtype='str')

3. 데이터 타입
순위                 int64
판매상품ID               str
상품명                  str
판매가                int64
저자                   str
출판사                  str
발행일       datetime64[us]
분야                   str
dtype: object

4. 컬럼별 결측치 개수
순위        0
판매상품ID    0
상품명       0
판매가       0
저자        0
출판사       0
발행일       0
분야        0
dtype: int64

5. 판매가 앞의 5개 값
0    13500
1    11700
2    19800
3    16200
4    12600
Name: 판매가, dtype: int64

6. 발행일 앞의 5개 값
0   2014-05-19
1   2013-04-01
2   2025-04-15
3   2025-03-28
4   2022-12-22
Name: 발행일, dtype: datetime64[us]

7. 판매상품ID 중복 개수
0

8. 데이터 앞의 5개 행
   순위         판매상품ID         상품명    판매가   저자   출판사        발행일     분야
0   1  S000000610612      소년이 온다  13500   한강    창비 2014-05-19     소설
1   2  S000001632467          모순  11700  양귀자    쓰다 2013-04-01     소설
2   3  S000216262757  결국 국민이 합니다  19800  이재명  오마이북 2025-04-15  정치/사회
3   4  S00

전처리가 완료된 df_books 데이터프레임을 대상으로 데이터 구조와 품질을 최종 확인했습니다.
- shape를 통해 전체 행과 열의 개수를 확인했습니다.
- columns를 통해 필요한 컬럼이 정상적으로 포함되어 있는지 확인했습니다.
- dtypes를 통해 각 컬럼이 목적에 맞는 자료형으로 변환되었는지 확인했습니다.
- isna().sum()을 통해 컬럼별 결측치 현황을 확인했습니다.
- 판매가와 발행일의 앞 5개 값을 확인하여 값과 형식이 정상적으로 저장되었는지 점검했습니다.
- 판매상품ID의 중복 개수를 확인했습니다.
- head()를 통해 전처리된 데이터의 앞부분을 최종적으로 확인했습니다.
위 검증을 통해 이후 분석에 사용할 df_books 데이터의 구조, 자료형, 결측치, 주요 값 및 중복 여부를 확인했습니다.

## 실습 11. 전처리 데이터 저장하기

다음 Chapter에서는 이 데이터를 다시 사용합니다.

따라서 전처리한 DataFrame을 CSV 파일로 저장합니다.

In [41]:
df_books.to_csv("book_bestseller_clean.csv", encoding="utf-8-sig", index=False)

## 12. 인공지능을 이용해 결과 Markdown 작성하기

실습을 완료하면 Notebook에 이번 작업을 설명하는 Markdown을 작성합니다.

1. 무엇을 확인했는지
원본 데이터의 크기와 필요한 컬럼 수를 확인했습니다. 또한 저자와 분야 컬럼의 결측치, 판매상품ID의 중복 여부, 판매가와 발행일의 자료형을 점검했습니다.
2. 어떤 전처리를 했는지
분석에 필요한 컬럼을 선택하고, 결측치와 중복 데이터를 확인했습니다. 판매가와 발행일 컬럼도 분석에 사용할 수 있도록 자료형을 정리했습니다.
3. 다음 분석을 위해 어떤 상태가 되었는지
전처리가 끝난 데이터는 도서 가격, 발행일, 저자, 분야 등을 기준으로 분석할 수 있는 형태가 되었습니다. 이후 분야별 도서 수, 가격 분포, 발행일별 추이 등의 분석에 활용할 수 있습니다.